README.md
# myca

Create my CA center, generate a self signed x509 certificate, issue server certificate from node.js via openssl. Multiple center supported. RSA, EC(P-256, P-384) supported.


[![GitHub tag](https://img.shields.io/github/tag/waitingsong/node-myca.svg)]()
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![](https://img.shields.io/badge/lang-TypeScript-blue.svg)]()
[![ci](https://github.com/waitingsong/node-myca/workflows/ci/badge.svg)](https://github.com/waitingsong/node-myca/actions?query=workflow%3A%22ci%22)
[![codecov](https://codecov.io/github/waitingsong/node-myca/branch/main/graph/badge.svg?token=wTaSMKz3Ne)](https://codecov.io/github/waitingsong/node-myca)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lernajs.io/)



## 安装全局依赖
```sh
npm i -g c8 lerna rollup tsx
```

## Installing
```bash
npm install --save myca
```

## CLI
- [myca-cli](https://www.npmjs.com/package/myca-cli)
- Installing by `npm i -g myca-cli`
- Command help
  - `myca`
  - `myca initca -h`
  - `myca issue -h`
- Example
  ```sh
  myca initca --days=10950 --alg=ec --pass=capass \
    --cn="Root CA" --ou="waitingsong.com" --o="waitingsong" --l="CD" --c=CN \
  ```

  ```sh
  myca issue --kind=server --days=3650 --pass=mypass \
    --cn="waitingsong.com" --o="waitingsong" --c=CN --caKeyPass=capass \
    --centerName=default --alg=ec \
    --ips="127.0.0.1, 192.168.0.1" \
    --SAN="localhost" 
  ```

  ```sh
  myca issue --kind=client --days=3650 --pass=mypass \
    --cn="client" --o="it" --c=CN --caKeyPass=capss \
    --centerName=default --alg=ec 
  ```


## Usage
- Initialize default center
  ```ts
  import { initDefaultCenter } from 'myca'

  await initDefaultCenter()
  ```


## Packages

| Package      | Version                |
| ------------ | ---------------------- |
| [`myca`]     | [![main-svg]][main-ch] |
| [`myca-cli`] | [![cli-svg]][cli-ch]   |



## License
[MIT](LICENSE)


### Languages
- [English](README.md)
- [中文](README.zh-CN.md)

<br>

[`myca`]: https://github.com/waitingsong/node-myca/tree/main/packages/myca
[main-svg]: https://img.shields.io/npm/v/myca.svg?maxAge=7200
[main-ch]: https://github.com/waitingsong/node-myca/tree/main/packages/myca/CHANGELOG.md

[`myca-cli`]: https://github.com/waitingsong/node-myca/tree/main/packages/myca-cli
[cli-svg]: https://img.shields.io/npm/v/myca-cli.svg?maxAge=7200
[cli-ch]: https://github.com/waitingsong/node-myca/tree/main/packages/myca-cli/CHANGELOG.md





CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [5.1.0](https://github.com/waitingsong/node-myca/compare/v5.0.0...v5.1.0) (2023-08-21)


### Features

* **cli:** use tsx instead of ts-node-esm as cli Hashbang ([c70d916](https://github.com/waitingsong/node-myca/commit/c70d91672d36806ac366cbbc05df357d30caf75b))





# [5.0.0](https://github.com/waitingsong/node-myca/compare/v4.2.0...v5.0.0) (2023-07-26)


### Features

* **myca:** genIssueSubj() not allow value of Array ([4c2d50e](https://github.com/waitingsong/node-myca/commit/4c2d50eee1cefdbffe7d2050c5108b9bb5f04fe0))
* **myca:** update types for ca cert generation ([083533a](https://github.com/waitingsong/node-myca/commit/083533afb302827d26eaf7c3815f1035ad31637e))





# [4.2.0](https://github.com/waitingsong/node-myca/compare/v4.1.7...v4.2.0) (2023-07-26)


### Features

* **myca-cli:** detect confusing command within parseCmd() ([c3018ba](https://github.com/waitingsong/node-myca/commit/c3018ba649a30072ac4dcb0eebbf99035cf29f9d))





## [4.1.7](https://github.com/waitingsong/node-myca/compare/v4.1.6...v4.1.7) (2023-07-25)

**Note:** Version bump only for package myca





## [4.1.6](https://github.com/waitingsong/node-myca/compare/v4.1.5...v4.1.6) (2023-07-24)


### Bug Fixes

* **myca:** normalize path ([170a2ba](https://github.com/waitingsong/node-myca/commit/170a2bae74fa5c99af270b527ee0337e094f0007))





## [4.1.5](https://github.com/waitingsong/node-myca/compare/v4.1.4...v4.1.5) (2023-07-24)

**Note:** Version bump only for package myca





## [4.1.4](https://github.com/waitingsong/node-myca/compare/v4.1.3...v4.1.4) (2023-07-24)


### Bug Fixes

* **myca-cli:** value of centerName and centerPath within initCenterCli() ([03bc699](https://github.com/waitingsong/node-myca/commit/03bc699c85ecf2397804426fb9e2833cf070ab85))





## [4.1.3](https://github.com/waitingsong/node-myca/compare/v4.1.2...v4.1.3) (2023-07-21)


### Bug Fixes

* **myca:** kind for client ([8ff64bd](https://github.com/waitingsong/node-myca/commit/8ff64bdef05dc994b92507787aa021f36f45d8b1))





## [4.1.2](https://github.com/waitingsong/node-myca/compare/v4.1.1...v4.1.2) (2023-07-21)


### Bug Fixes

* **myca:** genCert() not use kind during create csr ([1f594b3](https://github.com/waitingsong/node-myca/commit/1f594b3570ea2e3ad1e3b691ba971359cfadab32))
* **myca:** tmp files not removed ([f4a0d14](https://github.com/waitingsong/node-myca/commit/f4a0d14aae62f7485853f66916ecdb7a00909455))
* **myca:** tpl.conf.client ([e835bf2](https://github.com/waitingsong/node-myca/commit/e835bf250f2a00317ecb4cee20ad9ad6b5fbcd5b))





## [4.1.1](https://github.com/waitingsong/node-myca/compare/v4.1.0...v4.1.1) (2023-07-21)

**Note:** Version bump only for package myca





# [4.1.0](https://github.com/waitingsong/node-myca/compare/v4.0.4...v4.1.0) (2023-07-21)


### Bug Fixes

* **myca-cli:** parseCliArgs() argv not merged ([e77b028](https://github.com/waitingsong/node-myca/commit/e77b028a9881b725b19bc9e5d1a84924d47e5ae5))
* **myca:** options.caKeyPass may be number ([4475d57](https://github.com/waitingsong/node-myca/commit/4475d57a8d8561d51a12abdf0103b489d9a1b445))
* **myca:** trim serialContent ([ef73676](https://github.com/waitingsong/node-myca/commit/ef73676b9814991003fb5785e03ceaf69ea7498a))


### Features

* **myca-cli:** pass debug to issueCli() ([1b1fcde](https://github.com/waitingsong/node-myca/commit/1b1fcdebf4fe485e3b75333c99f35b71820d7743))
* **myca:** add debug parameter of functions ([99bc860](https://github.com/waitingsong/node-myca/commit/99bc8603674354a49ea7383476b5e5ece89e6996))
* **myca:** options of runOpenssl() accepts debug property ([77dd565](https://github.com/waitingsong/node-myca/commit/77dd5659f6b75888053ec66e3dd681a4921ac869))
* **myca:** set param conf of genCert() Partial ([eb26f7a](https://github.com/waitingsong/node-myca/commit/eb26f7abc09e5a8c5c31d14b75ae722fd6132b4e))
* **myca:** update tpl.conf.server ([8ca83fb](https://github.com/waitingsong/node-myca/commit/8ca83fb5fe20d3c45c87a2a7339701bce8532efb))





## [4.0.4](https://github.com/waitingsong/node-myca/compare/v4.0.3...v4.0.4) (2023-07-21)

**Note:** Version bump only for package npm-mono-base





## [4.0.3](https://github.com/waitingsong/node-myca/compare/v4.0.2...v4.0.3) (2023-07-20)


### Bug Fixes

* **myca:** runOpenssl() ([126a75e](https://github.com/waitingsong/node-myca/commit/126a75e6c0cd6ca8b99ec835a93e4b8fbb18c0cc))





## [4.0.2](https://github.com/waitingsong/node-myca/compare/v4.0.1...v4.0.2) (2023-07-20)

**Note:** Version bump only for package npm-mono-base





## [4.0.1](https://github.com/waitingsong/node-myca/compare/v4.0.0...v4.0.1) (2023-07-20)

**Note:** Version bump only for package npm-mono-base





# [4.0.0](https://github.com/waitingsong/node-myca/compare/v3.2.0...v4.0.0) (2023-07-20)


### Bug Fixes

* catch test error ([5efe156](https://github.com/waitingsong/node-myca/commit/5efe1565f9ab1766a7db9c4055bba570ea9eb49c))
* createDir() path resolve under linux ([c6d1274](https://github.com/waitingsong/node-myca/commit/c6d1274ef6be06e9862023401ad5bf95fc644c49))
* deps, peerDeps might empty ([e3ab52b](https://github.com/waitingsong/node-myca/commit/e3ab52ba14e483c9a49babaffbdbae6f7d77b41d))
* error TS1345: An expression of type 'void' cannot be tested for truthiness ([0085713](https://github.com/waitingsong/node-myca/commit/00857132333434009e302d78f2455ef09a8deaef))
* options not covered within createFile() ([a2ae4e8](https://github.com/waitingsong/node-myca/commit/a2ae4e826f9ed5c29f241fdff2df85928999b3ae))
* path require parse by normalize() within createDir() ([371a313](https://github.com/waitingsong/node-myca/commit/371a31358bd6375929db935abf74e9637f09adb3))
* revert ts-node to '5.0.1' ([cc83ade](https://github.com/waitingsong/node-myca/commit/cc83ade848f646ddf3913f2bfe430cae7756b26e))

test/20_init.test.ts
/// <reference types="mocha" />

import {
  basename,
  createDirAsync,
  createFileAsync,
  dirExists,
  isDirExists,
  isFileExists,
  join,
  promisify,
  tmpdir,
  unlinkAsync,
} from '@waiting/shared-core'
import * as assert from 'power-assert'
import * as rmdir from 'rimraf'
import { concat, forkJoin, iif, EMPTY } from 'rxjs'
import { catchError, concatMap, mergeMap, tap } from 'rxjs/operators'

import * as myca from '../src/index'
import { initialConfig } from '../src/lib/config'


const filename = basename(__filename)
const tmpDir = join(tmpdir(), 'myca-tmp')
const pathPrefix = 'myca-test-center'
const randomPathG = `${tmpDir}/${pathPrefix}-${Math.random()}`

// dir contains conf file and folders
// config.defaultCenterPath = normalize(`${config.userHome}/${config.centerDirName}`)
initialConfig.defaultCenterPath = `${randomPathG}/${initialConfig.centerDirName}`

describe(filename, () => {
  before(async () => {
    await createDirAsync(tmpDir)
  })
  after(done => {
    rmdir(tmpDir, err => err ? console.error(err) : done())
  })

  it('Should initDefaultCenter() works', done => {
    const centerPath$ = myca.initDefaultCenter().pipe(
      tap(path => {
        assert(
          path === initialConfig.defaultCenterPath,
          `result not expected. result: "${path}", expected: "${initialConfig.defaultCenterPath}"`,
        )
      }),
    )
    const exists$ = dirExists(initialConfig.defaultCenterPath).pipe(
      tap(path => {
        if (! path) {
          assert(false, `default center folder not exists, path: "${initialConfig.defaultCenterPath}"`)
        }
      }),
    )
    const centerInited$ = myca.isCenterInited('default').pipe(
      tap(inited => {
        assert(
          inited === true,
          `isCenterInited('default') says folder not exits. path: "${initialConfig.defaultCenterPath}"`,
        )
      }),
    )

    const first$ = centerPath$.pipe(
      concatMap(() => exists$),
      concatMap(() => centerInited$),
    )
    // initialize again
    const second$ = centerPath$.pipe(
      catchError(() => EMPTY),
      tap(() => {
        assert(false, 'should throw error during duplicate initialization, but NOT')
      }),
    )

    concat(
      first$,
      second$,
    ).subscribe(
      () => {},
      (err: Error) => assert(false, err.message),
      done,
    )

    // not rm for below test
    // rmdir(join(config.defaultCenterPath, '../'), (err) => err && console.error(err))
  })


  it('Should getCenterPath() works', done => {
    const centerPath$ = myca.getCenterPath('default')
    const exists$ = centerPath$.pipe(
      mergeMap(dirExists),
    )
    const isDefaultCenterInited$ = myca.isCenterInited('default')

    forkJoin(
      exists$,
      isDefaultCenterInited$,
    ).subscribe(
      ([path, inited]) => assert(path ? inited : ! inited),
      err => assert(false, err),
      done,
    )
  })



  // ------------------ at last

  it('Should getCenterPath() works with empty centerList', async () => {
    const file = join(initialConfig.defaultCenterPath, initialConfig.centerListName)

    try {
      if (await isFileExists(file)) {
        await unlinkAsync(file)
      }
      await createFileAsync(file, '')
      assert(await myca.getCenterPath('center').toPromise() === '', 'should return empty')
    }
    catch (ex) {
      assert(false, ex)
    }
  })


  it('Should initCenter() works without default Center', async () => {
    const random = Math.random()
    const centerName = `${pathPrefix}-${random}`
    const randomPath = `${tmpDir}/${pathPrefix}-${random}`
    const centerPath = `${randomPath}/${initialConfig.centerDirName}`
    const path = join(initialConfig.defaultCenterPath, '..')

    if (await isDirExists(path)) {
      const rmdirAsync = promisify(rmdir)

      try {
        await rmdirAsync(path)
      }
      catch (ex) {
        assert(false, `unlink default Center failed. path: "${path}"`)
      }
    }

    try {
      await myca.initCenter(centerName, centerPath).toPromise()
      assert(false, 'initCenter() should throw error, but NOT')
    }
    catch (ex) {
      assert(true)
    }

    assert(! await isDirExists(centerPath), `path should not exists: "${centerPath}"`)
  })

})



test/40_cert.test.ts
/// <reference types="node" />
/// <reference types="mocha" />

import { tmpdir } from 'os'
import { basename, join, normalize } from 'path'
import * as assert from 'power-assert'
import rewire = require('rewire')
import * as rmdir from 'rimraf'

import * as myca from '../src/index'
import { isDirExists, isFileExists } from '../src/lib/common'
import { config, initialCaOpts } from '../src/lib/config'


const filename = basename(__filename)
const tmpDir = tmpdir()
const random = Math.random()
const pathPrefix = 'myca-test-center'
const randomPath = `${tmpDir}/${pathPrefix}-${random}`
const mods = rewire('../src/lib/cert')

config.isWin32 = process.platform === 'win32' ? true : false
config.userHome = config.isWin32 ? normalize(process.env.USERPROFILE || '') : normalize(`${process.env.HOME}`)
config.defaultCenterPath = `${randomPath}/${config.centerDirName}`
config.openssl = normalize(config.openssl)


describe(filename, () => {
  beforeEach(async () => {
    try {
      await myca.initDefaultCenter()
    }
    catch (ex) {
      return
    }
  })
  after(() => {
    rmdir(randomPath, (err) => err && console.error(err))
  })


  it('Should initCaCert() works', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    try {
      await myca.initCaCert(opts)
    }
    catch (ex) {
      return assert(false, ex)
    }
  })

  it('Should initCaCert() works with invalid param', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
    }

    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with blank centerName', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.centerName = ''
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with fake centerName', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.centerName = 'fake'
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with blank pass', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.pass = ''
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with blank C', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.C = ''
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with blank CN', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.CN = ''
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with zero days', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.days = 0
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with negative days', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.days = -1
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with invalid alg', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.alg = <'rsa'> ''
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })

  it('Should initCaCert() works with invalid hash', async () => {
    const opts: myca.CaOpts = {
      ...initialCaOpts,
      days: 10950,
      pass: 'mycapass',
      CN: 'My Root CA',
      OU: 'waitingsong.com',
      C: 'CN',
    }

    opts.hash = <'sha256'> 'fake'
    try {
      await myca.initCaCert(opts)
      assert(false, 'initCaCert() should throw err, but NOT')
    }
    catch (ex) {
      assert(true)
    }
  })


})


src/index.ts
/**
 * node-myca
 *
 * @author waiting
 * @license MIT
 * @link https://github.com/waitingsong/node-myca
 */


import { isWin32, userHome } from '@waiting/shared-core'
import { normalize } from 'path'

import { initialConfig } from './lib/config'

/* istanbul ignore next */
initialConfig.appDir = __dirname + '/..'

/* istanbul ignore next */
initialConfig.isWin32 = isWin32
initialConfig.userHome = userHome
// dir contains conf file and folders
initialConfig.defaultCenterPath = normalize(`${initialConfig.userHome}/${initialConfig.centerDirName}`)
initialConfig.openssl = normalize(initialConfig.openssl)

/* istanbul ignore next */
if (! initialConfig.userHome) {
  throw new Error('path of user profile empty')
}

export * from './lib/center'
export * from './lib/cert'
export * from './lib/model'
export { initialCaOpts, initialCertOpts } from './lib/config'


src/lib/init.ts
import { exec } from 'child_process'
import { normalize } from 'path'

import { initOpensslConfig } from './center'
import { genCaCert, saveCaCrt } from './cert'
import {
  createDir,
  createFile,
  getCenterPath,
  getOpensslVer,
  isDirExists,
  isFileExists,
  updateCenterList } from './common'
import { config, initialCaOpts } from './config'
import { CaOpts, CertOpts } from './model'

config.isWin32 = process.platform === 'win32' ? true : false
config.userHome = config.isWin32 ? normalize(process.env.USERPROFILE || '') : normalize(`${process.env.HOME}`)
config.defaultCenterPath = normalize(`${config.userHome}/${config.centerDirName}`) // dir contains conf file and folders
config.openssl = normalize(config.openssl)

if ( ! config.opensslVer) {
  getOpensslVer(config.openssl).then(ver => {
    config.opensslVer = ver
  })
}
if ( ! config.userHome) {
  throw new Error('value of user profile (path) empty')
}

const folders: string[] = [config.dbDir, config.serverDir, config.clientDir, config.dbCertsDir]
const initialFiles: string[] = [config.centerListName]
export const userHome = config.userHome


// create defaultCenterPath and center folders/files
export async function initDefaultCenter(): Promise<void> {
  const centerName = 'default'

  if (await isCenterInited(centerName)) {
    return Promise.reject('default center initialized already')
  }

  await createDir(config.defaultCenterPath) // create default ca dir under userHome
  await createInitialFiles(config.defaultCenterPath, initialFiles)  // must before createCenter()
  await createCenter(centerName, config.defaultCenterPath)  // create default cneter dir under userHome
}


// create center path and folders/files
export async function initCenter(centerName: string, path: string): Promise<void> {
  if (centerName === 'default') {
    return Promise.reject('init default center by calling method of initDefaultCenter()')
  }
  if ( ! await isCenterInited('default')) {
    return Promise.reject('default center must be initialized first')
  }
  if (await isCenterInited(centerName)) {
    return Promise.reject(`center of "${centerName}" initialized already.`)
  }

  await createDir(path) // create default ca dir under userHome
  await createCenter(centerName, path)  // create default cneter dir under userHome
  console.log(`CenterPath name: ${centerName}, path: ${path}`)
}

export async function isCenterInited(centerName: string): Promise<boolean> {
  const centerPath = await getCenterPath(centerName)

  if ( ! centerPath) {
    return false
  }
  if (await isDirExists(centerPath)) {
    return true
  }
  return false
}

// create center dir to store output certifacates
export async function createCenter(centerName: string, path: string): Promise<void> {
  if ( ! await isCenterInited(centerName)) {
    await createDir(path)
  }
  for (let i = 0, len = folders.length; i < len; i++) {
    const dir = `${path}/${folders[i]}`

    if ( ! await isDirExists(dir)) {
      await createDir(dir)
    }
  }
  await initDbFiles(path)
  await updateCenterList(centerName, path)
  await initOpensslConfig(config.configName, path)
}


async function createInitialFiles(path: string, files: string[]): Promise<void> {
  for (let i = 0, len = files.length; i < len; i++) {
    const file = `${path}/${files[i]}`
    const data = { default: path }

    await createFile(file, data)
  }
}

async function initDbFiles(path: string): Promise<void> {
  const db = `${path}/${config.dbDir}`

  await createFile(`${db}/serial`, '01', { mode: 0o644 })
  await createFile(`${db}/index`, '', { mode: 0o644 })
  await createFile(`${db}/index.attr`, 'unique_subject = no', { mode: 0o644 })
}


export async function initCaCert(issueOpts: CaOpts): Promise<void> {
  const opts = <CertOpts> { ...initialCaOpts, ...issueOpts }

  if ( ! opts.centerName) {
    return Promise.reject('centerName empty')
  }
  if ( ! await isCenterInited(opts.centerName)) {
    return Promise.reject(`center: ${opts.centerName} not initialized yes`)
  }
  const centerPath = await getCenterPath(opts.centerName)
  const file = normalize(`${centerPath}/${config.caCrtName}`)

  if (await isFileExists(file)) {
    return Promise.reject(`CA file exists, should unlink it via unlinkCaCert(centerName). file: "${file}"`)
  }
  const certRet = await genCaCert(opts)

  await saveCaCrt(config, opts, certRet.cert)
}


src/lib/model.ts
// param options of fs.writeFile()
export interface WriteFileOptions {
  encoding?: string | null
  mode?: number
  flag?: string
}


/**
 * json object converted from center-list.json
 * {
 *  "default": "/home/mark/.myca"
 *  "app1": "opt/app1/.myca"
 *  "app2": "opt/app2/foo"
 * }
 */
export interface CenterList {
  default: string
  [name: string]: string
}

export interface Config {
  appDir: string  // base directory of this module
  dbDir: string
  dbCertsDir: string
  serverDir: string
  clientDir: string
  caKeyName: string // ca.key
  caCrtName: string // ca.crt

  userHome: string  // userprofile path
  centerDirName: string // .myca
  defaultCenterPath: string // userHome/centerDirName
  centerListName: string  // store in userHome/centerDirName/ , only one

  openssl: string   // openssl cli
  opensslVer: string  // dynamic calcu
  isOpensslCmdValid: boolean

  isWin32: boolean
  configName: string  // dynamic assign for win32
  confTpl: string // config template for win32
}

export type Alg = 'rsa' | 'ec'    // algorithm
export interface PrivateKeyOpts {
  // serial: string
  centerName: 'default' | string
  alg: Alg
  pass: string
  /** for alg==rsa */
  keyBits: number
  /** for alg==ec */
  ecParamgenCurve?: 'P-256' | 'P-384'
}

// passed by customer
export interface CaOpts {
  /** key name of log dir */
  centerName?: 'default' | string
  alg?: Alg
  days: number
  pass: string  // at least 4 chars
  keyBits?: number // for rsa
  ecParamgenCurve?: 'P-256' | 'P-384' // for alg==ec
  hash?: 'sha256' | 'sha384'

  /** Common Name */
  CN: string
  /** Organizational Unit Name */
  OU?: string
  /** Organization Name */
  O?: string
  /** Country Name (2 letter code) */
  C?: string
  /** State or Province Name */
  ST?: string
  /** Locality Name (eg, city) */
  L?: string
  emailAddress?: string
  [prop: string]: any
}

export type CertDNkeys = keyof CertDN

export interface CertDN {
  CN: string    // Common Name
  OU?: string    // Organizational Unit Name
  O?: string    // Organization Name
  C: string    // Country Name (2 letter code)
  ST?: string   // State or Province Name
  L?: string    // Locality Name (eg, city)
  emailAddress?: string
  SAN?: string[]  // subjectAltName
  ips?: string[] // subjectAltName ip
}

// sign csr
export interface SignOpts {
  kind: 'ca' | 'server' | 'client'
  centerPath: string  // default as config.defaultCenterPath
  days: number
  hash: 'sha256' | 'sha384'
  caCrtFile: string
  caKeyFile: string
  caKeyPass: string
  csrFile: string
  configFile?: string // openssl config file . default centerPath/.config
  SAN?: string[]  // subjectAltName DNS
  ips?: string[] // subjectAltName ip. NOT support ip mask such as 192.168.0.*
}

// passed by customer
export interface CertOpts extends CertDN {
  kind: 'ca' | 'server' | 'client'
  // serial?: string
  centerName?: 'default' | string  // key name of log dir
  alg?: Alg
  days: number
  pass: string // at least 4 chars
  caKeyPass: string
  keyBits?: number // for rsa
  ecParamgenCurve?: 'P-256' | 'P-384' // for alg==ec
  hash?: 'sha256' | 'sha384'
}

// inner usage
export interface IssueOpts extends SignOpts, CertDN {
  serial: string  // hex
  centerName: 'default' | string  // key name of center dir
  alg: Alg
  days: number
  pass: string
  keyBits: number // for rsa
  ecParamgenCurve: 'P-256' | 'P-384' // for alg==ec
  hash: 'sha256' | 'sha384'
}

export interface CertDetail {
  privateKey: string
  pubKey: string
  privateFile: string
  pubFile: string
}

export interface CertInfo {
  secure?: CertDetail   // available during issue with pass
  unsecure: CertDetail
}

export interface KeysRet {
  pubKey: string     // pubkey pem
  privateKey: string  // private key pem
  privateUnsecureKey: string  // private key pem without pass encrypted
  pass: string
  privateKeyFile: string
  privateUnsecureKeyFile: string
}

export interface IssueCaCertRet {
  centerName: string
  privateKey: string  // private key pem
  pass: string
  privateKeyFile: string
  cert: string  // certificate pem
  crtFile: string // certificate file path
}

export interface IssueCertRet extends KeysRet {
  centerName: string
  caKeyFile: string
  caCrtFile: string
  csr: string
  csrFile: string
  cert: string  // certificate pem
  crtFile: string // certificate file path
  pfxFile?: string  // for client
}

export interface InitialFile {
  name: string
  defaultValue: string | number
  mode?: number
}

export interface PfxOpts {
  privateKeyFile: string
  privateKeyPass?: string
  crtFile: string
  pfxPass: string
}

export interface StreamOpts {
  args: string[]
  runOpts: any
  rtpl: string
}


src/lib/config.ts
import {
  CaOpts,
  CertDNkeys,
  CertOpts,
  Config,
  InitialFile,
  IssueCaCertRet,
  IssueCertRet,
  IssueOpts,
  PrivateKeyOpts,
  SignOpts } from './model'


export const initialConfig: Config = {
  appDir: '',
  dbDir: 'db',
  dbCertsDir: 'db/certs',
  serverDir: 'server',
  clientDir: 'client',
  caKeyName: 'ca.key',
  caCrtName: 'ca.crt',

  userHome: '',
  centerDirName: '.myca',
  defaultCenterPath: '',
  centerListName: 'center-list.json',

  openssl: 'openssl',
  opensslVer: '',
  isOpensslCmdValid: false,

  isWin32: false,
  configName: '.config',  // under asset/
  confTpl: 'tpl.conf',  // under asset/
}

export const initialPrivateKeyOpts: PrivateKeyOpts = {
  // serial: 0,
  centerName: 'default',
  alg: 'rsa',
  pass: '',
  keyBits: 2048, // for alg==rsa
  ecParamgenCurve: 'P-256', // for alg==ec
}
export const initialCaOpts: CaOpts = {
  kind: 'ca',
  centerName: 'default',
  alg: 'rsa',
  days: 10950,  // 30years
  pass: '',
  keyBits: 4096,  // for alg==rsa
  ecParamgenCurve: 'P-256',
  hash: 'sha384',
  CN: '',
  OU: '',
  O: '',
  C: '',
  ST: '',
  L: '',
  emailAddress: '',
}

// sign csr
export const initialSignOpts: SignOpts = {
  kind: 'server',
  centerPath: '',
  days: 750,  // 2year
  hash: 'sha256',
  caCrtFile: '',
  caKeyFile: '',
  caKeyPass: '',
  csrFile: '',
  configFile: '',
  SAN: [],
  ips: [],
}

export const initialCertOpts: CertOpts = {
  // CertDN
  kind: 'server',
  days: 750,  // 2year
  hash: 'sha256',
  caKeyPass: '',
  SAN: [],
  ips: [],

  centerName: 'default',
  alg: 'rsa',
  pass: '',
  keyBits: 2048,
  ecParamgenCurve: 'P-256', // for alg==ec
  // DNs
  CN: '',
  OU: '',
  O: '',
  C: '',
  ST: '',
  L: '',
  emailAddress: '',
}

export const initialIssueOpts: IssueOpts = {
  serial: '', // hex
  centerName: 'default',
  alg: 'rsa',
  pass: '',
  keyBits: 2048,
  ecParamgenCurve: 'P-256', // for alg==ec
  // DNs
  CN: '',
  OU: '',
  O: '',
  C: '',
  ST: '',
  L: '',
  emailAddress: '',

  ...initialSignOpts,
}

export const reqSubjectFields: CertDNkeys[] = [
  'CN',
  'OU',
  'O',
  'L',
  'ST',
  'C',
  'emailAddress',
]

export const initialCaCertRet: IssueCaCertRet = {
  centerName: '',
  privateKey: '',
  pass: '',
  privateKeyFile: '',
  cert: '',  // certificate pem
  crtFile: '', // certificate file path
}

export const initialCertRet: IssueCertRet = {
  centerName: '',
  caKeyFile: '',
  caCrtFile: '',
  pubKey: '',     // pubkey pem
  privateKey: '',  // private key pem
  privateUnsecureKey: '',  // private key pem
  pass: '',
  privateKeyFile: '',
  privateUnsecureKeyFile: '',
  csr: '',
  csrFile: '',
  cert: '',  // certificate pem
  crtFile: '', // certificate file path
}

export const initialDbFiles: InitialFile[] = [
  {
    name: 'serial',
    defaultValue: '01',
    mode: 0o644,
  },
  {
    name: 'index',
    defaultValue: '',
    mode: 0o644,
  },
  {
    name: 'index.attr',
    defaultValue: 'unique_subject = no',
    mode: 0o644,
  },
]
